std/experimental/typecons.d

```diff
@@ -1074,3 +1074,258 @@ pure nothrow @system unittest
+ * `isAddress`)
+ *
+ * TODO ok to default `E` to `Exception`?
+ */
+struct Expected(T, E = Exception)
+if (!isInstanceOf!(Unexpected, T)) // an `Unexpected` cannot be `Expected` :)
+{
+    import std.algorithm.mutation : moveEmplace;
+
+    // TODO ok for default construction to initialize
+    // - _expectedValue = T.init (zeros)
+    // - _ok = true (better to have _isError so default is zero bits here aswell?)
+
+    /// Construct from expected value `expectedValue.`
+    this()(auto ref T expectedValue) @trusted
+    {
+        // TODO reuse opAssign?
+        moveEmplace(expectedValue, _expectedValue);
+        _ok = true;
+    }
+
+    /// Construct from unexpected value `unexpectedValue.`
+    this(Unexpected!E unexpectedValue) @trusted
+    {
+        // TODO reuse opAssign?
+        moveEmplace(unexpectedValue, _unexpectedValue);
+        _ok = false;
+    }
+
+    /// Assign from expected value `expectedValue.`
+    void opAssign()(auto ref T expectedValue) @trusted
+    {
+        // TODO is this ok?:
+        clear();
+        moveEmplace(expectedValue, _expectedValue);
+        _ok = true;
+    }
+
+    /// Assign from unexpected value `unexpectedValue.`
+    void opAssign(E unexpectedValue) @trusted
+    {
+        clear();
+        moveEmplace(unexpectedValue, _unexpectedValue);
+        _ok = false;
+    }
+
+    /// Clear (empty) contents.
+    private void clear() @trusted
+    {
+        release();
+        static if (isAddress!T)
+        {
+            _expectedValue = null;
+        }
+    }
+
+    /// Release any memory used to store contents.
+    private void release() @trusted
+    {
+        import std.traits : hasElaborateDestructor;
+        if (hasExpectedValue)
+        {
+            static if (!is(T == class))
+            {
+                static if (hasElaborateDestructor!T)
+                {
+                    destroy(_expectedValue);
+                }
+            }
+            _ok = false;
+        }
+        else
+        {
+            static if (!is(E == class))
+            {
+                static if (hasElaborateDestructor!E)
+                {
+                    destroy(_unexpectedValue);
+                }
+            }
+            destroy(_unexpectedValue);
+            // TODO change _ok?
+        }
+    }
+
+    /** Is `true` iff this has a expectedValue of type `T`. */
+    bool hasExpectedValue() const
+    {
+        return _ok;
+    }
+
+    import std.traits : CommonType;
+
+    /** Get current value if any or call function `elseWorkFun` with compatible return value.
+     *
+     * TODO is this anywhere near what we want?
+     */
+    CommonType!(T, typeof(elseWorkFun())) valueOr(alias elseWorkFun)() const
+    if (is(CommonType!(T, typeof(elseWorkFun()))))
+    {
+        if (hasExpectedValue)
+        {
+            return expectedValue;
+        }
+        else
+        {
+            return elseWorkFun(); // TODO is this correct
+        }
+    }
```

Comment:
Because of [issue 5710](https://issues.dlang.org/show_bug.cgi?id=5710), this method needs to be an external function that takes an explicit "self" parameter.

Reply:
Do you mean that `valueOr` needs to be an external function?

Reply:
Yes. Sorry for any confusion; this is my first time using Github's code review interface.

Reply:
Ok, thanks. What do you mean by explicit "self" parameter?

Reply:
Meaning it takes the `Expected` as a normal parameter instead of via the implicit `this` parameter:

```d
// reduced a bit for brevity
T valueOr(alias elseFun)(Expected!T self)
    if (is(typeof(elseFun()) == T))
{
    if (self.hasExpectedValue) return self.expectedValue;
    else return elseFun();
}
```

Reply:
Done.